### PR TITLE
8332858: References with escapes have broken positions after they are transformed

### PR DESCRIPTION
**Pull request by lahodaj**
If the javadoc comment contains a (Markdown) link like:
```
[java.util.Arrays#asList(Object\[\])]
```

The transformer that converts this link into the Javadoc link will not find the reference, as it is looking for `java.util.Arrays#asList(Object[])` (note the missing escapes), which is not present in the original text.

This patch tries to fix that by permitting optional escapes for all escapable character when searching for the reference, in case the literal search fails. This is done using regexp, although could presumably be done using a manual search.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8332858](https://bugs.openjdk.org/browse/JDK-8332858): References with escapes have broken positions after they are transformed (**Bug** - P3)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [b75f896f](https://git.openjdk.org/jdk/pull/19387/files/b75f896fb35f1c0c4a1b7216e1062b787a769757)
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19387/head:pull/19387` \
`$ git checkout pull/19387`

Update a local copy of the PR: \
`$ git checkout pull/19387` \
`$ git pull https://git.openjdk.org/jdk.git pull/19387/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19387`

View PR using the GUI difftool: \
`$ git pr show -t 19387`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19387.diff">https://git.openjdk.org/jdk/pull/19387.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19387#issuecomment-2129027116)

### CODE REVIEW

File: src/jdk.internal.md/share/classes/jdk/internal/markdown/MarkdownTransformer.java [hunk 1]
```diff
@@ -845,11 +849,31 @@ private int getRefPos(String ref, Link link) {
...
                 var index = s.lastIndexOf(ref);
                 if (index != -1) {
-                    return start + index;
+                    return new int[] {start + index, start + index + ref.length()};
+                } else {
+                    StringBuilder pattern = new StringBuilder(2 * ref.length());
+
+                    for (char c : ref.toCharArray()) {
+                        if (Escaping.ESCAPABLE.indexOf(c) >= 0) {
+                            pattern.append("\\\\?");
```
Comment (vicente-romero-oracle) [inline src/jdk.internal.md/share/classes/jdk/internal/markdown/MarkdownTransformer.java:858]: suggestion, dunno how critical is this code but regex usually are a tax on performance, I would consider not using them

Comment (lahodaj) [reply src/jdk.internal.md/share/classes/jdk/internal/markdown/MarkdownTransformer.java:858]: Sorry for the belated answer. Yes, regexps are usually not very performant, but it is only happening when the exact match fails. And, hopefully, the regexp should not be too difficult to handle. I was considering writing the search by hand, and I can, but it seems like a lot of code to handle a case like this. I can write the search manually, though.

Comment (vicente-romero-oracle) [reply src/jdk.internal.md/share/classes/jdk/internal/markdown/MarkdownTransformer.java:858]: ok if you think they won't be a problem in this case, I'm fine with it, we can always refactor the code if needed

Comment (jonathan-gibbons) [reply src/jdk.internal.md/share/classes/jdk/internal/markdown/MarkdownTransformer.java:858]: I think the use of regex is overkill, but maybe not inherently wrong. The only characters that need to be escaped, and _must_ be escaped are `\[\]`

>Unescaped square bracket characters are not allowed inside the opening and closing square brackets of [link labels](https://spec.commonmark.org/0.31.2/#link-label).

and, `[]` can only occur as such in adjacent pairs, so you ought to be able to get away with a simple `ref.replaceAll("[]", `\\[\\]")`

While Markdown may permit these of escapes for other characters in other kinds of links, for the extended form of links (that map to `{@link[plain]...}`) we specify the string must be the signature, with the only exception of needing to escape the square brackets.

Note in particular, this line, at about line 555 in this file.
```
                    var l = label.replace("\\[\\]", "[]");
```

**Review by jonathan-gibbons — CHANGES_REQUESTED** [review]

**Review by jonathan-gibbons — APPROVED** [review]

**Pull request closed without merging**